ntoskrnl/mm/ARM3/sysldr.c

```diff
@@ -268,28 +256,168 @@ MiLocateExportName(IN PVOID DllBase,
+                                                   &ExportSize);
+    if (!ExportDirectory)
+        return STATUS_INVALID_PARAMETER;
+
+    /* Setup name tables */
+    NameTable = (PULONG)((ULONG_PTR)DllBase +
+                         ExportDirectory->AddressOfNames);
+    OrdinalTable = (PUSHORT)((ULONG_PTR)DllBase +
+                             ExportDirectory->AddressOfNameOrdinals);
+
+    /* Get the ordinal */
+    Ordinal = NameToOrdinal(ExportName,
+                            DllBase,
+                            ExportDirectory->NumberOfNames,
+                            NameTable,
+                            OrdinalTable);
+
+    /* Check if we couldn't find it */
+    if (Ordinal == -1)
+        return STATUS_ORDINAL_NOT_FOUND;
+
+    /* Validate the ordinal */
+    if (Ordinal >= ExportDirectory->NumberOfFunctions)
+    {
+        /* It's not, fail */
+        return STATUS_ORDINAL_NOT_FOUND; // STATUS_PROCEDURE_NOT_FOUND;
+    }
 
     /* Resolve the address and write it */
     ExportTable = (PULONG)((ULONG_PTR)DllBase +
                            ExportDirectory->AddressOfFunctions);
-    Function = (PVOID)((ULONG_PTR)DllBase + ExportTable[Ordinal]);
+    FunctionAddress = (PVOID)((ULONG_PTR)DllBase + ExportTable[Ordinal]);
 
-    /* Check if the function is actually a forwarder */
-    if (((ULONG_PTR)Function > (ULONG_PTR)ExportDirectory) &&
-        ((ULONG_PTR)Function < ((ULONG_PTR)ExportDirectory + ExportSize)))
+    if (CheckIfForwarder)
     {
-        /* It is, fail */
-        return NULL;
+        /* Check if the function is actually a forwarder */
+        if (((ULONG_PTR)FunctionAddress > (ULONG_PTR)ExportDirectory) &&
+            ((ULONG_PTR)FunctionAddress < ((ULONG_PTR)ExportDirectory + ExportSize)))
+        {
+            ASSERT((FunctionAddress <= (PVOID)ExportDirectory) ||
+                   (FunctionAddress >= (PVOID)((ULONG_PTR)ExportDirectory + ExportSize)));
+
+            /* It is, fail */
+            return STATUS_PROCEDURE_NOT_FOUND;
+        }
     }
 
     /* We found it */
+    *Function = FunctionAddress;
+    return STATUS_SUCCESS;
+}
+
+/**
+ * @brief
+ * Finds the address in a loaded image of a given exported routine by name.
+ *
+ * @param[in]   DllBase
+ * The base address of the loaded image.
+ *
+ * @param[in]   ExportName
+ * The name of the export, given as an ANSI NULL-terminated string.
+ *
+ * @return
+ * The address of the named exported routine, or NULL if not found.
+ *
+ * @note
+ * This routine was originally named MiLocateExportName(), with a separate
+ * duplicated MiFindExportedRoutineByName() one (taking a PANSI_STRING)
+ * on Windows <= 2003. Both routines have been then merged and renamed
+ * to MiFindExportedRoutineByName() on Windows 8 (taking a PCSTR instead),
+ * and finally renamed and exported as RtlFindExportedRoutineByName() on
+ * Windows 10.
+ **/
+#define MiLocateExportName              RtlFindExportedRoutineByName
```

Comment:
Do we really need this? Why not simply rename those 3 instances?

Reply:
We could just rename the 3 instances. (I just wanted to keep their original names for documentation purposes, and to not add extra changes to the diff.)